src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -0,0 +1,508 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/util/af.h>
+
+#include <app/util/af-event.h>
+#include <app/util/attribute-storage.h>
+
+#include <app/CommandHandler.h>
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/attribute-type.h>
+#include <app/common/gen/attributes/Accessors.h>
+#include <app/common/gen/callback.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
+#include <app/common/gen/enums.h>
+#include <app/common/gen/ids/Attributes.h>
+#include <lib/core/CHIPEncoding.h>
+
+using namespace chip;
+using namespace chip::app::Clusters::Thermostat::Attributes;
+
+void emberAfThermostatClusterServerInitCallback(void)
+{
+    // TODO
+    // Get from the "real thermostat"
+    // current mode
+    // current occupied heating setpoint
+    // current unoccupied heating setpoint
+    // current occupied cooling setpoint
+    // current unoccupied cooling setpoint
+    // and update the zcl cluster values
+    // This should be a callback defined function
+    // with weak binding so that real thermostat
+    // can get the values.
+    // or should this just be the responsibility of the thermostat application?
+}
+
+/** @brief Thermostat Cluster Server Pre Attribute Changed
+ *
+ * Server Pre Attribute Changed
+ *
+ * @param endpoint      Endpoint that is being initialized
+ * @param attributeId   Attribute to be changed
+ * @param attributeType Attribute type
+ * @param size          Attribute size
+ * @param value         Attribute value
+ */
+EmberAfStatus emberAfThermostatClusterServerPreAttributeChangedCallback(chip::EndpointId endpoint, chip::AttributeId attributeId,
+                                                                        EmberAfAttributeType attributeType, uint16_t size,
+                                                                        uint8_t * value)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    int16_t Requested;
+
+    switch (attributeId)
+    {
+    case Ids::OccupiedHeatingSetpoint: {
+        int16_t AbsMinHeatSetpointLimit; // 7C (44.5 F) is the default
+        int16_t AbsMaxHeatSetpointLimit; // 30C (86 F) is the default
+        int16_t MinHeatSetpointLimit;    // 7C (44.5 F) is the default
+        int16_t MaxHeatSetpointLimit;    // 30C (86 F) is the default
+
+        status = GetAbsMinHeatSetpointLimit(endpoint, &AbsMinHeatSetpointLimit);
```

Comment:
I have been burned too many times by memory leaks created by multiple exits from functions.  Unless strictly required by Matter coding standards, I prefer to leave intact.

Reply:
Matter coding style is to prefer returns vs 'goto exit' because it creates easier to follow code and does not  cause undersirable 'variable declaration' sideffects (goto exit forces all variables C-style at the top).

In this case there  is no goto, however the code path complexity is higher than it could be. I would not block the PR on this, however would note that matter code style for memory leaks is to use RAII instead of C-style cleanup.